File: tests/mv-35/jmc-tests.js
#! /bin/env node
// we may need require to load re2
import { createRequire } from 'node:module';
const require = createRequire(import.meta.url);
import * as runtime from "json_model_runtime"
const JSON_MODEL_VERSION = "2";

const _jm_re_0_re = new runtime.RX("^[-\\w]*$", "s")
export var check_model_map = new Map()

function json_model_2(val, path, rep)
{
    return ((typeof val === 'string' || val instanceof String)) && runtime.jm_char_length(val) > 0
}

function json_model_3(val, path, rep)
{
    let res = Array.isArray(val) && val.length == 2
    if (res)
    {
        res = (typeof val[0] === 'boolean' || val[0] instanceof Boolean)
        if (res)
            res = true
    }
    return res
}

const _jm_re_0 = (s) => _jm_re_0_re.exec(s) !== null

function json_model_4(val, path, rep)
{
    let res = Array.isArray(val) && val.length == 3
    if (res)
    {
        res = (typeof val[0] === 'boolean' || val[0] instanceof Boolean)
        if (res)
        {
            res = ((typeof val[1] === 'string' || val[1] instanceof String)) && _jm_re_0(val[1], null, null)
            if (res)
                res = true
        }
    }
    return res
}

function json_model_5(val, path, rep)
{
    return json_model_2(val, null, null) || json_model_3(val, null, null) || json_model_4(val, null, null)
}

function json_model_1(val, path, rep)
{
    let res = Array.isArray(val)
    if (res)
    {
        for (let arr_0_idx = 0; arr_0_idx < val.length; arr_0_idx++)
        {
            let arr_0_item = val[arr_0_idx]
            res = json_model_5(arr_0_item, null, null)
            if (! res)
                break
        }
    }
    return res
}


var initialized = false

// differed module initializations
export function check_model_init()
{
    if (! initialized)
    {
        initialized = true;
        runtime.jm_set_rx(RegExp)
        check_model_map.set("", json_model_1)
        check_model_map.set("Comment", json_model_2)
        check_model_map.set("Test2", json_model_3)
        check_model_map.set("Test3", json_model_4)
        check_model_map.set("Test", json_model_5)
    }
}

// differed module cleanup
export function check_model_free()
{
    if (initialized)
        initialized = false;
}

export function check_model(val, name, rep)
{
    let checker = check_model_map.get(name)
    if (checker === undefined)
        throw `no checker for "${name}"`
    let path = rep !== null ? [] : null

    return checker(val, path, rep)
}

// possibly run as main based on a guess
import main from "json_model_runtime/main.js"

if (import.meta.url.endsWith(process.argv[1]))
    main(check_model_init, check_model, check_model_free)

File: tests/mv-35/jmc-tests.java

import json_model.*;
import java.util.Map;
import java.util.Set;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.regex.Pattern;
import java.util.regex.Matcher;

@SuppressWarnings("unchecked")
public class jmc_tests extends ModelChecker
{
    static public final String VERSION = "2";

    public Pattern _jm_re_0_pat = null;
    public Map<String, Checker> jmc_tests_map_pmap;

    public boolean json_model_2(Object val, Path path, Report rep)
    {
        return json.isString(val) && json.asString(val).startsWith("#");
    }

    public boolean json_model_3(Object val, Path path, Report rep)
    {
        boolean res = json.isArray(val) && json.arrayLength(val) == 2;
        if (res)
        {
            res = json.isBoolean(json.arrayItem(val, 0));
            if (res)
                res = true;
        }
        return res;
    }

    public boolean _jm_re_0(String val, Path path, Report rep)
    {
        return _jm_re_0_pat.matcher(val).find();
    }

    public boolean json_model_4(Object val, Path path, Report rep)
    {
        boolean res = json.isArray(val) && json.arrayLength(val) == 3;
        if (res)
        {
            res = json.isBoolean(json.arrayItem(val, 0));
            if (res)
            {
                res = json.isString(json.arrayItem(val, 1)) && _jm_re_0(json.asString(json.arrayItem(val, 1)), null, null);
                if (res)
                    res = true;
            }
        }
        return res;
    }

    public boolean json_model_5(Object val, Path path, Report rep)
    {
        return json_model_2(val, null, null) || json_model_3(val, null, null) || json_model_4(val, null, null);
    }

    public boolean json_model_1(Object val, Path path, Report rep)
    {
        boolean res = json.isArray(val);
        if (res)
        {
            int arr_0_idx = -1;
            Iterator<Object> arr_0_item_loop = json.arrayIterator(val);
            while (arr_0_item_loop.hasNext())
            {
                arr_0_idx++;
                Object arr_0_item = arr_0_item_loop.next();
                res = json_model_5(arr_0_item, null, null);
                if (! res)
                    break;
            }
        }
        return res;
    }


    public void init(JSON json)
    {
        if (!initialized)
        {
            try {
            _jm_re_0_pat = Pattern.compile("(?s)^[-\\w]*$");
            jmc_tests_map_pmap = new HashMap<String, Checker>();
            jmc_tests_map_pmap.put("", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_1(o, p, r);} });
            jmc_tests_map_pmap.put("Comment", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_2(o, p, r);} });
            jmc_tests_map_pmap.put("Test2", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_3(o, p, r);} });
            jmc_tests_map_pmap.put("Test3", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_4(o, p, r);} });
            jmc_tests_map_pmap.put("Test", new Checker() { public boolean call(Object o, Path p, Report r) { return json_model_5(o, p, r);} });
                super.init(json);
            }
            catch (Exception e) {
                throw new Error("cannot initialized model checker: " + e);
            }
        }
    }

    public void free()
    {
        if (initialized)
        {
            super.free();
            _jm_re_0_pat = null;
            jmc_tests_map_pmap = null;
        }
    }

    public Checker get(String name)
    {
        return jmc_tests_map_pmap.get(name);
    }

    public Set<String> models()
    {
        return jmc_tests_map_pmap.keySet();
    }

    static public void main(String[] args) throws Exception
    {
        ModelChecker checker = new jmc_tests();
        Main.main("jmc_tests", checker, VERSION, args);
    }
}
